### ISSUE TITLE
Error if you set Id as int

### ISSUE THREAD

**Issue by stesvis** [codecutout/JsonApiSerializer#97]
I tried to run the serialization example, but I changed all the `public string Id { get; set; }` to `public int Id { get; set; }` (safe to assume that most classes that have an Id property it's gonna be an integer.

```
var author = new Person
{
	Id = 9,
	FirstName = "Dan",
	LastName = "Gebhardt",
	Twitter = "dgeb",
};

var articles = new[] {
	new Article
	{
		Id = 1,
		Title = "JSON API paints my bikeshed!",
		Author = author,
		Comments = new List<Comment>
		{
			new Comment
			{
				Id = 5,
				Body = "First!",
				Author = new Person
				{
					Id = 2
				},
			},
			new Comment
			{
				Id = 12,
				Body = "I like XML better",
				Author = author,
			}
		}
	}
};
```

When I execute `string json = JsonConvert.SerializeObject(articles, new JsonApiSerializerSettings());` I get this exception:

> {System.InvalidCastException: Specified cast is not valid.
>   at (wrapper castclass) System.Object.__castclass_with_cache(object,intptr,intptr)
>   at JsonApiSerializer.Util.WriterUtil.ShouldWriteProperty[T] (System.Object value, Newtonsoft.Json.Serialization.JsonProperty prop, Newtonsoft.Json.JsonSerializer serializer, T& propValue) [0x00019] in <b13f188f534342b29336826ce418c62e>:0 
>   at JsonApiSerializer.JsonConverters.ResourceObjectConverter.WriteJson (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.JsonSerializer serializer) [0x0007a] in <b13f188f534342b29336826ce418c62e>:0 
>   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeConvertable (Newtonsoft.Json.JsonWriter writer, Newtonsoft.Json.JsonConverter converter, System.Object value, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonContainerContract collectionContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x00081] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs:650 
>   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonContract valueContract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x00071] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs:166 
>   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00079] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs:95 
>   at Newtonsoft.Json.Serialization.JsonSerializerProxy.SerializeInternal (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type rootType) [0x00008] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerProxy.cs:281 
>   at Newtonsoft.Json.JsonSerializer.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value) [0x00000] in /_/Src/Newtonsoft.Json/JsonSerializer.cs:1076 
>   at JsonApiSerializer.JsonConverters.ResourceObjectListConverter.WriteJson (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.JsonSerializer serializer) [0x00073] in <b13f188f534342b29336826ce418c62e>:0 
>   at JsonApiSerializer.JsonConverters.DocumentRootConverter.WriteJson (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.JsonSerializer serializer) [0x00127] in <b13f188f534342b29336826ce418c62e>:0 
>   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeConvertable (Newtonsoft.Json.JsonWriter writer, Newtonsoft.Json.JsonConverter converter, System.Object value, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonContainerContract collectionContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x00081] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs:650 
>   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonContract valueContract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x00071] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs:166 
>   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00079] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs:95 
>   at Newtonsoft.Json.Serialization.JsonSerializerProxy.SerializeInternal (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type rootType) [0x00008] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerProxy.cs:281 
>   at Newtonsoft.Json.JsonSerializer.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value) [0x00000] in /_/Src/Newtonsoft.Json/JsonSerializer.cs:1076 
>   at JsonApiSerializer.JsonConverters.DocumentRootConverter.ResolveAsRootData (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.JsonSerializer serializer) [0x00064] in <b13f188f534342b29336826ce418c62e>:0 
>   at JsonApiSerializer.JsonConverters.ResourceObjectListConverter.WriteJson (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.JsonSerializer serializer) [0x0000d] in <b13f188f534342b29336826ce418c62e>:0 
>   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeConvertable (Newtonsoft.Json.JsonWriter writer, Newtonsoft.Json.JsonConverter converter, System.Object value, Newtonsoft.Json.Serialization.JsonContract contract, Newtonsoft.Json.Serialization.JsonContainerContract collectionContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x00081] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs:650 
>   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.SerializeValue (Newtonsoft.Json.JsonWriter writer, System.Object value, Newtonsoft.Json.Serialization.JsonContract valueContract, Newtonsoft.Json.Serialization.JsonProperty member, Newtonsoft.Json.Serialization.JsonContainerContract containerContract, Newtonsoft.Json.Serialization.JsonProperty containerProperty) [0x00071] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs:166 
>   at Newtonsoft.Json.Serialization.JsonSerializerInternalWriter.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00079] in /_/Src/Newtonsoft.Json/Serialization/JsonSerializerInternalWriter.cs:95 
>   at Newtonsoft.Json.JsonSerializer.SerializeInternal (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x0023a] in /_/Src/Newtonsoft.Json/JsonSerializer.cs:1149 
>   at Newtonsoft.Json.JsonSerializer.Serialize (Newtonsoft.Json.JsonWriter jsonWriter, System.Object value, System.Type objectType) [0x00000] in /_/Src/Newtonsoft.Json/JsonSerializer.cs:1049 
>   at Newtonsoft.Json.JsonConvert.SerializeObjectInternal (System.Object value, System.Type type, Newtonsoft.Json.JsonSerializer jsonSerializer) [0x00028] in /_/Src/Newtonsoft.Json/JsonConvert.cs:664 
>   at Newtonsoft.Json.JsonConvert.SerializeObject (System.Object value, System.Type type, Newtonsoft.Json.JsonSerializerSettings settings) [0x00007] in /_/Src/Newtonsoft.Json/JsonConvert.cs:613 
>   at Newtonsoft.Json.JsonConvert.SerializeObject (System.Object value, Newtonsoft.Json.JsonSerializerSettings settings) [0x00000] in /_/Src/Newtonsoft.Json/JsonConvert.cs:591 
>   at VeloGuideApp.ViewModels.VelolisterApiPageViewModel.ExecuteSerializeCommand () [0x000c9] in R:\VeloGuide\App\VeloGuideApp\VeloGuideApp\ViewModels\_Examples\VelolisterApiPageViewModel.cs:100 }

What is the problem?

**Comment by stesvis** [comment]
All my classes have 
`public int Id { get; set; }` instead of `public string Id { get; set; }`
but when I **deserialize** the json response it correctly converts the string id to int.

Is there a way to make it work the same on **serialize**?
so if your Id is an int it would convert it to string?

**Comment by alex-davies** [comment]
Generally i'm a strong believer in that request/response models should be entirely different from internal models in which case you could just specify your Id correctly as a string.

However because of the amount of times this trips people up and the fact Json.net already does some deserialization magic to auto convert ints to strings, It might be easier to to auto convert ints to strings when serializing.

Feel free to raise a PR, or if you can wait a week I can have a fix in for this

**Comment by stesvis** [comment]
Thanks @alex-davies I can wait for sure.

For now I implemented this dirty trick :)

```
public class BaseDTO
{
	[JsonIgnore]
	public int Id
	{
		get { return int.Parse(sId); }
		set { sId = value.ToString(); }
	}

	/// <summary>
	/// json:api only works if id is a string
	/// </summary>
	[JsonProperty("id")]
	internal string sId { get; set; }
}
```

Then each of my DTOs extends this base class. When I serialize and deserialize I can use Id as int and it gets converted to string automatically for the serializer.

**Comment by alex-davies** [comment]
@stesvis this has now been fixed in 1.7.0 (https://github.com/codecutout/JsonApiSerializer/pull/98) you are free to use `int`, `long` or `Guid` as ids and they will be toString'ed when serializing

I have also made the error message a bit clearer when you have an Id of an unsupported type

**Comment by stesvis** [comment]
Awesome! I am sure this will help many! Great job!
